From d59ced5f666092a97a70316cc4ec36f14ab0e91e Mon Sep 17 00:00:00 2001
From: Cristy <urban-warrior@imagemagick.org>
Date: Thu, 6 Jul 2017 19:20:38 -0400
Subject: [PATCH] Heap buffer overflow in ReadOneMNGImage

a crafted file will cause x_off[i] out-of-bound operation vulnerability.

bug-debian: https://bugs.debian.org/870106
bug: https://github.com/ImageMagick/ImageMagick/issues/542
origin: https://github.com/ImageMagick/ImageMagick/commit/78d4c5db50fbab0b4beb69c46c6167f2c6513dec
---
 coders/png.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/coders/png.c b/coders/png.c
index 321dd35ae..981e6ffc4 100644
--- a/coders/png.c
+++ b/coders/png.c
@@ -5743,6 +5743,8 @@ static Image *ReadOneMNGImage(MngInfo* mng_info, const ImageInfo *image_info,
 
               for (i=(ssize_t) first_object; i <= (ssize_t) last_object; i++)
               {
+                if ((i < 0) || (i >= MNG_MAX_OBJECTS))
+                  continue;
                 if (mng_info->exists[i] && !mng_info->frozen[i] &&
                     (p-chunk) < (ssize_t) (length-8))
                   {
